### PR TITLE
#167164990 Comments And Replies On Novels

### PR DESCRIPTION
**Pull request by Ukhu**
#### Pivotal tracker story
[#167164990](https://www.pivotaltracker.com/story/show/167164990)

#### What does this PR do?
   - Creates an endpoint for posting comments
   - Creates an endpoint for replying comments

#### Summary of Task
   - Design database structure and a holistic approach to implementing the feature
   - Think about possible edge cases and write tests for each instance
   - Come up with the algorithm for posting and replying comments in single-level threads
   - Create and add the necessary middlewares, helpers, and services
   - Create `postComment` endpoint
   - Create `replyComment` endpoint

#### How can this be tested?
   - pull the branch to your local repository: `git pull origin/feature/167164990/comment-on-novels`
   - install packages and dependencies: `npm install`
   - run automated tests: `npm test`
   - test manually on Postman: `/api/v1/novels/:slug/comments` AND `/api/v1/novels/:slug/comments/:id`
   - supply values for the field `commentBody` in the body of the request
   - generate jwt tokens using the same secretKey as in your `.env` with a payload of the `id` of a seeded user in the database or a random `id` of a user not in the database and add to the `authorization` header of the request
   - happy testing

#### Screenshots (if appropriate)
Screenshot 2019-08-02 at 4 28 17 AM

### CODE REVIEW

File: src/middlewares/verifyToken.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+dotenv.config();
+
+const { SECRET_KEY } = process.env;
+
+const { responseMessage } = helpers;
+const { userServices: { findUser } } = services;
+
+export default (request, response, next) => {
+  const token = request.headers.authorization || request.query.token;
```
Comment (benisi) [inline src/middlewares/verifyToken.js:21]: Great work!

I will like to suggest that you search for Bearer prepend in your token and take it off i.e "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NiwiaWF0IjoxNTY0ODI3MzI1LCJleHAiOjE1NjU0MzIxMjV9.p1yXbgzLUPGtF73o5p-ASLqnYVnTX4efUECtGlawtN4" This is standard and it helps to differentiate different types of a token in a larger application. What do you think?

Comment (Ukhu) [reply src/middlewares/verifyToken.js:21]: Yeah I understand your point, and I appreciate your feedback. However, I considered the context of our product, and like we discussed this morning, I feel it won't be necessary to prepend the token with 'Bearer' and still go ahead to remove it in the code. I feel it's more straightforward this way.

Comment (benisi) [reply src/middlewares/verifyToken.js:21]: ok,

Comment (benisi) [reply src/middlewares/verifyToken.js:21]: LGTM

**Review by benisi — APPROVED** [review]

File: src/controllers/commentController.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ * @returns {json} - json
+ */
+const replyComment = async (request, response) => {
+  const { commentBody } = request.body;
+  const { slug, parentId: parentCommentId } = request.params;
+  const {
+    id: userId, firstName, lastName, isSubscribed
+  } = request.user;
+  try {
+    if (!isSubscribed) return responseMessage(response, 403, { error: 'you have to be subscribed to reply a comment' });
```
Comment (joel-ace) [inline src/controllers/commentController.js]: Can you add the curly braces to your if statements?

File: src/controllers/commentController.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    const novel = await findNovel(slug);
+    if (!novel) return responseMessage(response, 404, { error: 'novel not found' });
+    const parentComment = await findComment(parentCommentId);
+    if (!parentComment) return responseMessage(response, 404, { error: 'no comment found with the provided id' });
+    if (!parentComment.novelId) return responseMessage(response, 403, { error: 'you can only reply a parent comment' });
+    const novelId = null;
```
Comment (joel-ace) [inline src/controllers/commentController.js:70]: why are we setting `novelId` to null here?

Comment (Ukhu) [reply src/controllers/commentController.js:70]: This is how I differentiated a direct comment on a novel and a reply to a comment. For replies, the `novelId` is set to null, while for direct comments the `parentId` is set to null.

**Review by Nerocodes — APPROVED** [review]
LGTM
